src/wp-includes/class-wp-html-tag-processor.php

```diff
@@ -0,0 +1,2047 @@
+	 * text offsets of that token may shift; the bookmark is
+	 * kept updated with those shifts and remains stable unless
+	 * the entire span of text in which the token sits is removed.
+	 *
+	 * Release bookmarks when they are no longer needed.
+	 *
+	 * Example:
+	 * ```
+	 *     <main><h2>Surprising fact you may not know!</h2></main>
+	 *           ^  ^
+	 *            \-|-- this `H2` opener bookmark tracks the token
+	 *
+	 *     <main class="clickbait"><h2>Surprising fact you may no…
+	 *                             ^  ^
+	 *                              \-|-- it shifts with edits
+	 * ```
+	 *
+	 * Bookmarks provide the ability to seek to a previously-scanned
+	 * place in the HTML document. This avoids the need to re-scan
+	 * the entire thing.
+	 *
+	 * Example:
+	 * ```
+	 *     <ul><li>One</li><li>Two</li><li>Three</li></ul>
+	 *                                 ^^^^
+	 *                                 want to note this last item
+	 *
+	 *     $p = new WP_HTML_Tag_Processor( $html );
+	 *     $in_list = false;
+	 *     while ( $p->next_tag( [ 'tag_closers' => $in_list ? 'visit' : 'skip' ] ) ) {
+	 *         if ( 'UL' === $p->get_tag() ) {
+	 *             if ( $p->is_tag_closer() ) {
+	 *                 $in_list = false;
+	 *                 $p->set_bookmark( 'resume' );
+	 *                 if ( $p->seek( 'last-li' ) ) {
+	 *                     $p->add_class( 'last-li' );
+	 *                 }
+	 *                 $p->seek( 'resume' );
+	 *                 $p->release_bookmark( 'last-li' );
+	 *                 $p->release_bookmark( 'resume' );
+	 *             } else {
+	 *                 $in_list = true;
+	 *             }
+	 *         }
+	 *
+	 *         if ( 'LI' === $p->get_tag() ) {
+	 *             $p->set_bookmark( 'last-li' );
+	 *         }
+	 *     }
+	 * ```
+	 *
+	 * Because bookmarks maintain their position they don't
+	 * expose any internal offsets for the HTML document
+	 * and can't be used with normal string functions.
+	 *
+	 * Because bookmarks allocate memory and require processing
+	 * for every applied update they are limited and require
+	 * a name. They should not be created inside a loop.
+	 *
+	 * Bookmarks are a powerful tool to enable complicated behavior;
+	 * consider double-checking that you need this tool if you are
+	 * reaching for it, as inappropriate use could lead to broken
+	 * HTML structure or unwanted processing overhead.
+	 *
+	 * @param string $name Identifies this particular bookmark.
+	 * @return false|void
+	 * @throws Exception Throws on invalid bookmark name if WP_DEBUG set.
+	 */
+	public function set_bookmark( $name ) {
+		if ( null === $this->tag_name_starts_at ) {
+			return false;
+		}
+
+		if ( ! array_key_exists( $name, $this->bookmarks ) && count( $this->bookmarks ) >= self::MAX_BOOKMARKS ) {
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+				throw new Exception( "Tried to jump to a non-existent HTML bookmark {$name}." );
```

Comment:
we had a fair amount of internal discussion on this while building it out. the goal was to make improper use fail as soon as possible when devs are building it, but avoid breaking any sites in production. will a warning accomplish that? I wasn't familiar with the idioms in WordPress to do that.

we want to make sure we don't accidentally hide these failures because they indicate someone has written their code in a way that should never be done; they've written code guaranteed to fail.

Reply:
Thanks for the insight @dmsnell!

I'm not totally sure of the best approach here. Here's some usage in Core for our reference that may help:

Excluding external libraries, there are ~8 instances of `throw new Exception()` in Core. The rest seem to be `trigger_error()` (~46 instances I believe).

For most, they use the default error level of `E_USER_NOTICE` (20 instances).

Other cases have:
- `WP_DEBUG ? E_USER_WARNING : E_USER_NOTICE` (7)
- `E_USER_DEPRECATED` (18 instances - not relevant here but mentioning for completeness)
- `E_USER_WARNING` (1)

Reply:
@costdev I think `trigger_error( "…", E_USER_WARNING )` will be appropriate when `WP_DEBUG` is set. On one hand I'd still prefer execution to halt, because these are _only_ setup for situations that are guaranteed to fail/evidence of a software bug and not a conditional circumstance. On the other hand I guess you could be running a site and a bug from someone else's code in a plugin crashes your dev setup which would be unnecessarily frustrating.

There's probably no need to lesson the severity and `E_USER_NOTICE` doesn't really apply based on its description as "could also happen in the normal course of running a script."

`E_USER_DEPRECATED` isn't relevant either as nothing is replicated.

Does this all seem fair to you?

Reply:
triggered user warnings in d8fdf41c11